File: internal/repo/staging_test.go
package repo

import (
	"strings"
	"testing"
)

func TestBuildHunkPatch_AddLines(t *testing.T) {
	hunk := Hunk{
		Header: "@@ -0,0 +1,2 @@",
		Lines: []HunkLine{
			{Type: "add", Content: "first line", NewLine: 1},
			{Type: "add", Content: "second line", NewLine: 2},
		},
	}
	patch := BuildHunkPatch("foo.go", hunk)

	if !strings.Contains(patch, "diff --git a/foo.go b/foo.go") {
		t.Error("patch missing diff --git header")
	}
	if !strings.Contains(patch, "--- a/foo.go") {
		t.Error("patch missing --- header")
	}
	if !strings.Contains(patch, "+++ b/foo.go") {
		t.Error("patch missing +++ header")
	}
	if !strings.Contains(patch, "@@ -0,0 +1,2 @@") {
		t.Error("patch missing hunk header")
	}
	if !strings.Contains(patch, "+first line") {
		t.Error("patch missing +first line")
	}
	if !strings.Contains(patch, "+second line") {
		t.Error("patch missing +second line")
	}
}

func TestBuildHunkPatch_DeleteLines(t *testing.T) {
	hunk := Hunk{
		Header: "@@ -1,2 +0,0 @@",
		Lines: []HunkLine{
			{Type: "del", Content: "remove me", OldLine: 1},
		},
	}
	patch := BuildHunkPatch("bar.go", hunk)

	if !strings.Contains(patch, "-remove me") {
		t.Error("patch missing -remove me")
	}
	if strings.Contains(patch, "+remove me") {
		t.Error("patch should not contain +remove me")
	}
}

func TestBuildHunkPatch_ContextLines(t *testing.T) {
	hunk := Hunk{
		Header: "@@ -1,3 +1,3 @@",
		Lines: []HunkLine{
			{Type: "context", Content: "before", OldLine: 1, NewLine: 1},
			{Type: "del", Content: "old", OldLine: 2},
			{Type: "add", Content: "new", NewLine: 2},
			{Type: "context", Content: "after", OldLine: 3, NewLine: 3},
		},
	}
	patch := BuildHunkPatch("baz.go", hunk)

	if !strings.Contains(patch, " before") {
		t.Error("context line should start with space")
	}
	if !strings.Contains(patch, "-old") {
		t.Error("del line should start with -")
	}
	if !strings.Contains(patch, "+new") {
		t.Error("add line should start with +")
	}
	if !strings.Contains(patch, " after") {
		t.Error("context line should start with space")
	}
}

func TestBuildHunkPatch_EndsWithNewline(t *testing.T) {
	hunk := Hunk{
		Header: "@@ -1,1 +1,1 @@",
		Lines:  []HunkLine{{Type: "add", Content: "x", NewLine: 1}},
	}
	patch := BuildHunkPatch("f.go", hunk)

	if !strings.HasSuffix(patch, "\n") {
		t.Error("patch should end with newline (required by git apply)")
	}
}

func TestBuildHunkPatch_PathWithSubdir(t *testing.T) {
	hunk := Hunk{
		Header: "@@ -1,1 +1,1 @@",
		Lines:  []HunkLine{{Type: "add", Content: "x", NewLine: 1}},
	}
	patch := BuildHunkPatch("internal/repo/foo.go", hunk)

	if !strings.Contains(patch, "diff --git a/internal/repo/foo.go b/internal/repo/foo.go") {
		t.Error("patch should contain the full sub-directory path")
	}
}

File: frontend/src/stores/staging.test.ts
import { describe, it, expect } from 'vitest';
import { buildHunkPatch, isConflictedFile } from './staging';
import type { FileStatus } from './staging';
import { repo } from '../../wailsjs/go/models';

function makeHunk(
  header: string,
  lines: { type: string; content: string; oldLine: number; newLine: number }[]
): repo.Hunk {
  return repo.Hunk.createFrom({ header, lines });
}

describe('buildHunkPatch', () => {
  it('produces the correct patch format for add lines', () => {
    const hunk = makeHunk('@@ -0,0 +1,2 @@', [
      { type: 'add', content: 'first line', oldLine: 0, newLine: 1 },
      { type: 'add', content: 'second line', oldLine: 0, newLine: 2 },
    ]);
    const patch = buildHunkPatch('foo.go', hunk);

    expect(patch).toContain('diff --git a/foo.go b/foo.go');
    expect(patch).toContain('--- a/foo.go');
    expect(patch).toContain('+++ b/foo.go');
    expect(patch).toContain('@@ -0,0 +1,2 @@');
    expect(patch).toContain('+first line');
    expect(patch).toContain('+second line');
  });

  it('prefixes del lines with -', () => {
    const hunk = makeHunk('@@ -1,1 +0,0 @@', [
      { type: 'del', content: 'remove me', oldLine: 1, newLine: 0 },
    ]);
    const patch = buildHunkPatch('bar.go', hunk);

    expect(patch).toContain('-remove me');
    expect(patch).not.toContain('+remove me');
  });

  it('prefixes context lines with a space', () => {
    const hunk = makeHunk('@@ -1,3 +1,3 @@', [
      { type: 'context', content: 'before', oldLine: 1, newLine: 1 },
      { type: 'del', content: 'old', oldLine: 2, newLine: 0 },
      { type: 'add', content: 'new', oldLine: 0, newLine: 2 },
      { type: 'context', content: 'after', oldLine: 3, newLine: 3 },
    ]);
    const patch = buildHunkPatch('baz.go', hunk);

    expect(patch).toContain(' before');
    expect(patch).toContain(' after');
  });

  it('ends with a newline (required by git apply)', () => {
    const hunk = makeHunk('@@ -1,1 +1,1 @@', [
      { type: 'add', content: 'x', oldLine: 0, newLine: 1 },
    ]);
    const patch = buildHunkPatch('f.go', hunk);
    expect(patch.endsWith('\n')).toBe(true);
  });

  it('matches the Go BuildHunkPatch output for a mixed hunk', () => {
    const hunk = makeHunk('@@ -1,2 +1,2 @@', [
      { type: 'context', content: 'ctx', oldLine: 1, newLine: 1 },
      { type: 'del', content: 'old', oldLine: 2, newLine: 0 },
      { type: 'add', content: 'new', oldLine: 0, newLine: 2 },
    ]);
    const patch = buildHunkPatch('file.go', hunk);
    const expected = [
      'diff --git a/file.go b/file.go',
      '--- a/file.go',
      '+++ b/file.go',
      '@@ -1,2 +1,2 @@',
      ' ctx',
      '-old',
      '+new',
      '',
    ].join('\n');

    expect(patch).toBe(expected);
  });
});

describe('isConflictedFile', () => {
  const file = (staged: string, unstaged: string): FileStatus =>
    ({ staged, unstaged, path: 'f', oldPath: '' }) as FileStatus;

  it('returns true when staged is U (both modified conflict)', () => {
    expect(isConflictedFile(file('U', 'U'))).toBe(true);
  });

  it('returns true when staged is U and unstaged is something else', () => {
    expect(isConflictedFile(file('U', ' '))).toBe(true);
  });

  it('returns true when unstaged is U', () => {
    expect(isConflictedFile(file(' ', 'U'))).toBe(true);
  });

  it('returns true for AA (both added) conflict', () => {
    expect(isConflictedFile(file('A', 'A'))).toBe(true);
  });

  it('returns true for DD (both deleted) conflict', () => {
    expect(isConflictedFile(file('D', 'D'))).toBe(true);
  });

  it('returns false for normal staged-only modification', () => {
    expect(isConflictedFile(file('M', ' '))).toBe(false);
  });

  it('returns false for normal unstaged modification', () => {
    expect(isConflictedFile(file(' ', 'M'))).toBe(false);
  });

  it('returns false for untracked files', () => {
    expect(isConflictedFile(file('?', '?'))).toBe(false);
  });

  it('returns false for staged new file (A + space)', () => {
    expect(isConflictedFile(file('A', ' '))).toBe(false);
  });
});
